### DISCUSSION TITLE
why using dict.copy() is unreliable to detect ORM changes in JSON structures; better ways to mutate JSON structures?

### DISCUSSION THREAD

**Discussion by PiotrCzapla** [sqlalchemy/sqlalchemy#6866] [Usage Questions] [answered]
### Describe the bug

In order to trigger an update statement on a json field I'm using the trick with assigning a copy of a dictionary to the json field. Unfortunately this fails to issue the update statement when deferred field is already fetched.

The code below shows that the object is not modified when the deferred json field is fetched before the field is reassigned. 
The only way to trigger the update is to add a new key to a copy of the json field, on the first level. If a nested field is updated the update statement wont be issued.

### To Reproduce

```python
## connection URI as first arg
import sys, time
import sqlalchemy as sa
from sqlalchemy import orm

URI = sys.argv[1] # 'postgresql+psycopg2://...' 

Base = orm.declarative_base()
class Cache(Base):
    __tablename__='cache'
    id = sa.Column(sa.Integer, primary_key=True)
    counter = sa.Column(sa.Integer)
    obj_ = orm.deferred(sa.Column('obj_', sa.JSON))


engine = sa.create_engine(URI, future=True)

Base.metadata.create_all(engine)
session = orm.sessionmaker(bind=engine)()

session.execute(sa.delete(Cache).where(Cache.id == 1))
session.commit()

# add a cached object 
obj={'id':1, 'tmp':1}
cobj = Cache(id=1, counter=0, obj_=obj)
cobj = session.merge(cobj)
print('is modified after merge', session.is_modified(cobj))   # True
session.commit()

cached_obj = session.get(Cache, 1)

# This will fail to issue the update statement
val = cached_obj.obj_.copy() 
cached_obj.obj_ = val
print(f'is modified: {session.is_modified(cached_obj)} after: cached_obj.obj_.copy()')
session.commit()

# This will fail to issue the update statement
cached_obj.obj_ # only if this line is present
val = obj.copy()
cached_obj.obj_ = val
print(f'is modified: {session.is_modified(cached_obj)} after: obj.copy()')
session.commit()

# Adding a new field is a work around that works  
val = {**cached_obj.obj_, 'ts':time.time()}
cached_obj.obj_ = val
print(f"is modified: {session.is_modified(cached_obj)} after: {{**cached_obj.obj_, 'ts':time.time()}}")
session.commit()

# However simply adding a new key to a fetched field wont work
cached_obj.obj_['ts'] = time.time()
print(f"is modified: {session.is_modified(cached_obj)} after: .obj_['ts'] = time.time()")
session.commit()
```

### Error

there is no error messages, the output of the script above looks as follows:
```
is modified: False after: cached_obj.obj_.copy()
is modified: False after: obj.copy()
is modified: True after: {**cached_obj.obj_, 'ts':time.time()}
is modified: False after: .obj_['ts'] = time.time()
is modified: True after: merge
```

### Versions

- OS: macos 11.5.1
- Python: 3.8.5
- SQLAlchemy:  1.4.22
- Database: postgresql
- DBAPI: postgresql+psycopg2

**Comment by zzzeek** [comment, marked as answer]
hi im not sure what you're trying to do.   the example shows no net change to the field, so there is no UPDATE statement.  also relying on .copy() is unreliable if there are mutable structures within the dictionary.  to ensure an UPDATE takes place no matter what, use [flag_modified](https://docs.sqlalchemy.org/en/14/orm/session_api.html?highlight=flag_modified#sqlalchemy.orm.attributes.flag_modified), which is just what the [sqlalchemy.ext.mutable](https://docs.sqlalchemy.org/en/14/orm/extensions/mutable.html?highlight=mutable#module-sqlalchemy.ext.mutable) does, which is what you really should use for this kind of case.


emits update statement:

```py
val = cached_obj.obj_.copy()
cached_obj.obj_ = val
orm.attributes.flag_modified(cached_obj, "obj_")
print(
    f"is modified: {session.is_modified(cached_obj)} after: cached_obj.obj_.copy()"
)
session.commit()

```

**Reply by PiotrCzapla** [reply]
Hi @zzzeek,  this is a minimal example, of an issue with a caching layer that we have on production.  My bad that I focused on technical inconsistency instead of a business issue. Our objects have nested dictionaries and when only such nested dicts are changed, the change is not detected and objects are not saved in to the database. 

This is problematic because the code appears to work flawlessly as most of the time there is an additional change to a field in the parent dictionary, so the objects are stored in db without issues.

Let me give you an example it might be easier to explain.

**Reply by zzzeek** [reply]
> Hi @zzzeek, this is a minimal example, of an issue with a caching layer that we have on production. My bad that I focused on technical inconsistency instead of a business issue. Our objects have nested dictionaries and when only such nested dicts are changed, the change is not detected and objects are not saved in to the database.

this is expected, when you modify a nested structure, that's the same structure in your original dictionary, so no net change is detected.    you would instead need to use `copy.deepcopy()` for your technique to work.  however, using `flag_modified()` is much cheaper and should work in all cases.

> 
> This is problematic because the code appears to work flawlessly as most of the time there is an additional change to a field in the parent dictionary, so the objects are stored in db without issues.

this is due to your not-recommended workaround of using copy() to trick the ORM into detecting a change.

**Reply by zzzeek** [reply]
below is a suite you can run with pytest that shows everything that's going on.   what it does not include is the detail of why "deferred" in your original example makes a difference, which is because when the deferred attribute is not loaded, the ORM saves on the SELECT by just assuming the value changed.  when the old value is present, it then can do the compare and see no net change.

```py
import copy

from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import inspect
from sqlalchemy import Integer
from sqlalchemy import JSON
from sqlalchemy import select
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import Session
from sqlalchemy.orm.attributes import flag_modified
Base = declarative_base()


class A(Base):
    __tablename__ = 'a'

    id = Column(Integer, primary_key=True)
    data = Column(JSON)

e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)

s = Session(e)

s.add(A(data={"foo": ["one", "two"]}))
s.commit()


def assert_no_update():
    s.flush()

    assert s.scalar(select(A.data)) == {"foo": ["one", "two"]}

    s.rollback()

def assert_update():
    s.flush()

    assert s.scalar(select(A.data)) == {"foo": ["one", "two", "three"]}

    s.rollback()

def test_one():
    a1 = s.query(A).first()

    a1.data["foo"].append("three")

    assert_no_update()

def test_two():
    a1 = s.query(A).first()

    a1.data = a1.data.copy()
    a1.data["foo"].append("three")

    # we copied, so it's a different dictionary:
    assert inspect(a1).committed_state["data"] is not a1.__dict__["data"]

    # but why no update?   because the embedded "foo" value is the same list:
    assert inspect(a1).committed_state["data"]["foo"] is a1.__dict__["data"]["foo"]

    # history does this:
    we_should_emit_an_update = inspect(a1).committed_state["data"]["foo"] != a1.__dict__["data"]["foo"]

    # and it's false
    assert not we_should_emit_an_update

    # shorthand way to see this, use get_history or the inspect.attrs.data.history
    new_value, existing_value, old_value = inspect(a1).attrs.data.history

    # no new value or old value
    assert not new_value
    assert not old_value

    # existing_value is it
    assert existing_value == [{"foo": ["one", "two", "three"]}]

    # so no update
    assert_no_update()


def test_three():
    a1 = s.query(A).first()

    a1.data = copy.deepcopy(a1.data)
    a1.data["foo"].append("three")

    # now with deepcopy, everything that happened in test_two is different,
    # history now has a new and deleted value, no existing value


    assert inspect(a1).committed_state["data"] is not a1.__dict__["data"]

    # inner list is different
    assert inspect(a1).committed_state["data"]["foo"] is not a1.__dict__["data"]["foo"]

    # they compare differently
    we_should_emit_an_update = inspect(a1).committed_state["data"]["foo"] != a1.__dict__["data"]["foo"]
    assert we_should_emit_an_update


    # history shows it
    new_value, existing_value, old_value = inspect(a1).attrs.data.history

    assert not existing_value
    assert old_value == [{"foo": ["one", "two"]}]
    assert new_value == [{"foo": ["one", "two", "three"]}]

    # update happens
    assert_update()


def test_four():
    a1 = s.query(A).first()

    a1.data["foo"].append("three")

    # shorthand to all of that above, "flag_modified" just forces the
    # "old" value to a constant so that UPDATE happens unconditionally
    flag_modified(a1, "data")
    assert_update()
```

**Reply by PiotrCzapla** [reply]
I see now that I've extracted wrong minimal example. The code I've prepared does not show the issue I'm facing. As my production code makes deep copies  (the objects are deserialised from json). Yet in some cases it fails to save the changes. So the issue is somewhere else. Give me till Monday to track it, and then we close this thread. 

Thank you for the workaround with `flag_modified()` .

**Reply by PiotrCzapla** [reply]
Actually once I find the issue I will start new discussion. Thank you for the help!

**Reply by zzzeek** [reply]
it seems really likely that it's some version of the same thing above, if you are only using copy().   change your use of copy() to deepcopy() and that will likely fix it.

**Comment by PiotrCzapla** [comment]
@zzzeek, I've got bitten again due to my lousy code, but It might be worth bringing up as there is an easy fix. 
I limited the use of json fields to flat structures so that a changes are detected when I make a  shallow copy `dict(a1.data)`.
Unfortunately this only works if the shallow copy is done before the modification. Otherwise we are making a modification to the cached dict object and no changes are detected. Here is an example that extends the above unit test. 



```python
def assert_rename():
    s.flush()

    assert s.scalar(select(A.data)) == {"bar": ["one", "two"]}

    s.rollback()

def test_broken_rename():
    a1 = s.query(A).first()

    a1.data["bar"] = a1.data['foo']
    del a1.data['foo']
    a1.data = dict(a1.data)               # this does not help as the cached object is already modified

    assert_no_update()

def test_working_rename():
    a1 = s.query(A).first()

    a1.data = dict(a1.data)
    a1.data["bar"] = a1.data['foo']
    del a1.data['foo']
    
    assert_rename()
```

I know that the code is poorly written as I should create a separate variable to make the copy more explicit, but I felt into the trap of think about the copy as a way to inform the library about about the change, and I've forgotten that the dict is cached somewhere and used later to detect changes.  This might be easy to fix by making the returned dict  immutable using 'types.MappingProxyType' or something similar with a better error message.  It won't solve all issues but it should let the user know that she really should not modify the field directly.  To make it work for all cases we would need to make a nested dicts and lists immutable too.

I'm worried about the backward compatibility here,  as I've seen the code that threat tuples different from lists, but I'm sure you have a process for that. Do you think such fix make sense?  If so can make a PR.

**Reply by zzzeek** [reply]
not sure if im following.  are you suggesting that SQLAlchemy's JSON datatype make it so that the returned JSON structures use immutable dicts/tuples?     I would consider this out of scope for SQLAlhcemy itself as I'm not familiar with an efficient process for deserializing JSON into an immutable structure.  if you wanted to implement this for your own code I would advise supplying a custom json deserializer: https://docs.sqlalchemy.org/en/14/core/engines.html?highlight=json_deserializer#sqlalchemy.create_engine.params.json_deserializer

**Reply by PiotrCzapla** [reply]
I've though that it is aligned with a general developer friendliness, something similar to the effort you make to warn about cross joins. But if it is out of scope, I understand. Thank you for the pointer to the docs, we will implement that on our side then.

**Reply by zzzeek** [reply]
we are developer friendly where it's practical but changing all JSON to immutable across the board would be a very drastic and complex change, that is unlikely to really be feasible in any case due to all the backwards incompatibility issues it would introduce.
